### DISCUSSION TITLE
How to disable Inlay Hints?

### DISCUSSION THREAD

**Discussion by xevrem** [joaotavora/eglot#1191] [Q&A] [answered]
inlays are nice... when you want them... when you don't they are annoying. they add a performance cost to everything, clutter the UI, and make traversing the buffer tricky resulting in unexpected jumps or requiring extra mental effort to properly parse the buffer. they can, and often do cause undesired line overflow and for complex types can be longer than the line content itself, etc... i.e., having them always-on is not the best idea.

as far as i can tell, inlay hints are enabled automagically, and there is no option to disable them other than to explicitly toggle them off via calling the mode function itself. Even further, it doesn't seem like there is safe way to disable it via a hook either as the only option that works "works" by calling the mode-off switch when the hook is called... and that is bad due to: "infinite" recursion as the hook is called on activation AND deactivation...

if i am wrong about the above, can someone please post a safe way to automatically disable it, because having to turn it off for each and every buffer that uses it becomes extremely annoying

**Comment by joaotavora** [comment, marked as answer]
Customize `eglot-ignored-server-capabilities` and add `:inlayHintProvider` there.

**Reply by joaotavora** [reply]
But the hook idea should also work. What infinite recursion are you talking about?

**Reply by xevrem** [reply]
thanks, this worked!

to answer your next question, since the hook is fired when enabling -and- disabling the inlay hints, turning it off right after it turned on, will fire the hook again... which calls it again, and then again, and then again... essentially causing it to recurse until the call-depth is exceeded. Testing the state of the mode (i.e., is this on or off), didn't seem to work for preventing it either. 🤷‍♀️

**Reply by joaotavora** [reply]
Where exactly and how did you add the `eglot-inlay-hints-mode` call?  I was taking about something as this: does this recurse infinitely?? https://www.reddit.com/r/emacs/comments/11yn3ba/comment/jd8u5kz/?utm_source=share&utm_medium=web2x&context=3

**Reply by xevrem** [reply]
i only wanted to have it fire if it was enabled, so i had it tied the minor-mode's hook 😅 
i.e., `(add-hook 'eglot-inlay-hints-mode-hook (lambda () (eglot-inlay-hints-mode -1)))`

i completely forgot that `eglot-managed-mode` hook was a thing

**Reply by joaotavora** [reply]
Oh....  adding a mode to its own mode hook is  a pretty good way to get infinite recursion :smile:
